### PR TITLE
Feat: Add `relabel_configs` field to AlertmanagerEndpoints 

### PR DESCRIPTION
**Pull request by mviswanathsai**
## Description
We don't support the relabel_configs per AM, along with ProxyConfig, FollowRedirects among other fields. We would like to extend support to these missing fields.

Partially closes: #6461 
## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.
Unit tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```Add relabel_configs field per AM

```

### CODE REVIEW

**Comment by mviswanathsai** [comment]
Should I also include proxy url configs in this PR? and maybe also `FollowRedirects` ? @simonpasquier

**Comment by simonpasquier** [comment]
no let's tackle one issue at a time.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 1]
```diff
@@ -1655,6 +1655,11 @@ type AlertmanagerEndpoints struct {
 	//
 	// +optional
 	EnableHttp2 *bool `json:"enableHttp2,omitempty"`
+
+	// List of alertmanager relabel configs
+	//
+	// +optional
+	RelabelConfigs []*RelabelConfig `json:"relabelConfigs,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1662]:
```suggestion
	RelabelConfigs []RelabelConfig `json:"relabelings,omitempty"`
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -1655,6 +1655,11 @@ type AlertmanagerEndpoints struct {
 	//
 	// +optional
 	EnableHttp2 *bool `json:"enableHttp2,omitempty"`
+
+	// List of alertmanager relabel configs
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1659]:
```suggestion
	// Relabel configuration applied to the discovered Alertmanagers.
```

File: pkg/prometheus/operator.go [hunk 3]
```diff
@@ -146,7 +146,7 @@ func ValidateRemoteWriteSpec(spec monitoringv1.RemoteWriteSpec) error {
 	return nil
 }
 
-func ValidateAlertmanagerEndpoints(am monitoringv1.AlertmanagerEndpoints) error {
+func ValidateAlertmanagerEndpoints(am monitoringv1.AlertmanagerEndpoints, p monitoringv1.PrometheusInterface) error {
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:149]: I commented elsewhere that it could be moved to pkg/prometheus/server. Can you do it in another PR that we'll merge first?

File: pkg/prometheus/operator.go [hunk 4]
```diff
@@ -169,6 +169,13 @@ func ValidateAlertmanagerEndpoints(am monitoringv1.AlertmanagerEndpoints) error
 		return fmt.Errorf("%s can't be set at the same time, at most one of them must be defined", strings.Join(nonNilFields, " and "))
 	}
 
+	if len(am.RelabelConfigs) != 0 {
+		err := validateRelabelConfigs(p, am.RelabelConfigs)
+		if err != nil {
+			return err
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:175]:
```suggestion
			return fmt.Errorf("invalid relabelings": err)
```

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -1839,6 +1839,10 @@ func (cg *ConfigGenerator) generateAlertmanagerConfig(alerting *monitoringv1.Ale
 			})
 		}
 
+		if len(am.RelabelConfigs) != 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1842]: no need to test the len

File: pkg/prometheus/operator.go [hunk 6]
```diff
@@ -169,6 +169,13 @@ func ValidateAlertmanagerEndpoints(am monitoringv1.AlertmanagerEndpoints) error
 		return fmt.Errorf("%s can't be set at the same time, at most one of them must be defined", strings.Join(nonNilFields, " and "))
 	}
 
+	if len(am.RelabelConfigs) != 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:172]: no need to test the len

File: pkg/prometheus/store.go [hunk 7]
```diff
@@ -70,9 +70,9 @@ func AddRemoteReadsToStore(ctx context.Context, store *assets.Store, namespace s
 	return nil
 }
 
-func AddAlertmanagerEndpointsToStore(ctx context.Context, store *assets.Store, namespace string, ams []monv1.AlertmanagerEndpoints) error {
+func AddAlertmanagerEndpointsToStore(ctx context.Context, store *assets.Store, namespace string, ams []monv1.AlertmanagerEndpoints, p monv1.PrometheusInterface) error {
```
Comment (simonpasquier) [inline pkg/prometheus/store.go:73]: same remark here, there's no need for this function to exist in pkg/prometheus

File: pkg/prometheus/operator.go (hunk 3 above)
Comment (mviswanathsai) [reply pkg/prometheus/operator.go:149]: Sure, I will make another PR for that change.

Comment (mviswanathsai) [reply pkg/prometheus/operator.go:149]: Would moving it to pkg/prometheus/server/operator.go be appropriate?

File: pkg/prometheus/store.go (hunk 7 above)
Comment (mviswanathsai) [reply pkg/prometheus/store.go:73]: Where should we put it? there is no pkg/prometheus/server/store.go. Should it be in  pkg/prometheus/server/operator.go again?

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1/prometheus_types.go:1662]: I think we can change the type after we merge #6479?

File: pkg/prometheus/operator.go (hunk 3 above)
Comment (ArthurSens) [reply pkg/prometheus/operator.go:149]: Sounds good to me!

File: pkg/prometheus/store.go (hunk 7 above)
Comment (ArthurSens) [reply pkg/prometheus/store.go:73]: I don't see a problem in creating a `store.go` file 🤔

**Comment by ArthurSens** [comment]
I'm not sure if I'm missing something, but are we adding two different relabeling options under alertmanager endpoints? I was almost sure there was only one relabeling option for alertmanagers

File: pkg/prometheus/operator.go (hunk 3 above)
Comment (mviswanathsai) [reply pkg/prometheus/operator.go:149]: The change has been made in #6480. Could you review?

**Comment by mviswanathsai** [comment]
Yes, as of v2.51.0, we have the ability to have relabeling rules for alerts per alertmanager: `alert_relabel_configs` in addition to the existing `relabel_configs` for alertmanager discovery.

File: pkg/prometheus/store.go (hunk 7 above)
Comment (mviswanathsai) [reply pkg/prometheus/store.go:73]: Ok then, I will create a `store.go` file and add this function there.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 8]
```diff
@@ -1671,6 +1671,11 @@ type AlertmanagerEndpoints struct {
 	//
 	// +optional
 	EnableHttp2 *bool `json:"enableHttp2,omitempty"`
+
+	// Relabel configuration applied to the discovered Alertmanagers.
+	//
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:1676]: Scratch that! I mixed up things with #6450 :)

**Review by simonpasquier — APPROVED** [review]

**Comment by simonpasquier** [comment]
Thanks!
